test/mainnet-fork/Uniswapv4.t.sol

```diff
@@ -839,6 +863,154 @@ contract MainnetControllerUniswapV4Tests is ForkTestBase {
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setMaxSlippage(address(uint160(uint256(_POOL_ID))), 0.98e18);
+        rateLimits.setRateLimitData(_SWAP_LIMIT_KEY, 2_000_000e18, 0);
+        vm.stopPrank();
+
+        deal(address(usdc), address(almProxy), 1_000_000e6);
+
+        vm.prank(relayer);
+        vm.expectRevert("MC/amountOutMin-too-low");
+        mainnetController.swapUniswapV4(_POOL_ID, address(usdc), 1_000_000e6, 980_000e6 - 1);
+
+        vm.prank(relayer);
+        mainnetController.swapUniswapV4(_POOL_ID, address(usdc), 1_000_000e6, 980_000e6);
+    }
+
+    function test_swapUniswapV4_revertsWhenInputTokenNotForPool() external {
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setMaxSlippage(address(uint160(uint256(_POOL_ID))), 0.98e18);
+        rateLimits.setRateLimitData(_SWAP_LIMIT_KEY, 2_000_000e18, 0);
+        vm.stopPrank();
+
+        vm.prank(relayer);
+        vm.expectRevert(IPoolManagerLike.CurrencyNotSettled.selector);
+        mainnetController.swapUniswapV4(_POOL_ID, address(dai), 1_000_000e6, 1_000_000e6);
+    }
+
+    function test_swapUniswapV4_revertsWhenAmountOutMinNotMet() external {
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setMaxSlippage(address(uint160(uint256(_POOL_ID))), 0.98e18);
+        rateLimits.setRateLimitData(_SWAP_LIMIT_KEY, 2_000_000e18, 0);
+        vm.stopPrank();
+
+        deal(address(usdc), address(almProxy), 1_000_000e6);
+
+        vm.prank(relayer);
+
+        // V4TooLittleReceived
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                IV4RouterLike.V4TooLittleReceived.selector,
+                999_280.652247e6 + 1,
+                999_280.652247e6
+            )
+        );
+
+        mainnetController.swapUniswapV4(_POOL_ID, address(usdc), 1_000_000e6, 999_280.652247e6 + 1);
+
+        vm.prank(relayer);
+        mainnetController.swapUniswapV4(_POOL_ID, address(usdc), 1_000_000e6, 999_280.652247e6);
+    }
+
+    function test_swapUniswapV4_revertsWhenRateLimitExceeded() external {
+        uint256 expectedDecrease = 1_000_000e18;
+
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setMaxSlippage(address(uint160(uint256(_POOL_ID))), 0.98e18);
+        rateLimits.setRateLimitData(_SWAP_LIMIT_KEY, expectedDecrease - 1, 0);
+        vm.stopPrank();
+
+        uint128 amountOutMin = _getSwapAmountOutMin(address(usdc), 1_000_000e6, 0.99e18);
+
+        deal(address(usdc), address(almProxy), 1_000_000e6);
+
+        vm.prank(relayer);
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        mainnetController.swapUniswapV4({
+            poolId       : _POOL_ID,
+            tokenIn      : address(usdc),
+            amountIn     : 1_000_000e6,
+            amountOutMin : amountOutMin
+        });
+
+        vm.prank(SPARK_PROXY);
+        rateLimits.setRateLimitData(_SWAP_LIMIT_KEY, expectedDecrease, 0);
+
```

Comment:
In this test lets do it by setting rate limit to 1m then swapping 1m + 1 and failing, then swapping 1m and passing